File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/SendFriendRequest.java
package uk.co.victoriajanedavis.chatapp.domain.interactors;

import javax.inject.Inject;

import io.reactivex.Single;
import io.reactivex.annotations.NonNull;
import uk.co.victoriajanedavis.chatapp.data.repositories.SentFriendRequestRepository;
import uk.co.victoriajanedavis.chatapp.domain.entities.FriendshipEntity;
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.SendInteractor;

public class SendFriendRequest implements SendInteractor<String, FriendshipEntity> {

    @NonNull
    private final SentFriendRequestRepository repository;


    @Inject
    public SendFriendRequest(@NonNull final SentFriendRequestRepository repository) {
        this.repository = repository;
    }

    @NonNull
    @Override
    public Single<FriendshipEntity> getSingle(@NonNull String username) {
        return repository.sendFriendRequest(username);
    }
}

File: app/src/main/java/uk/co/victoriajanedavis/chatapp/domain/interactors/SendFriendRequest.kt
package uk.co.victoriajanedavis.chatapp.domain.interactors

import io.reactivex.Completable
import javax.inject.Inject

import uk.co.victoriajanedavis.chatapp.data.repositories.SentFriendRequestRepository
import uk.co.victoriajanedavis.chatapp.domain.interactors.ReactiveInteractor.ActionInteractor

class SendFriendRequest @Inject constructor(
    private val repository: SentFriendRequestRepository
) : ActionInteractor<SendFriendRequest.FriendRequestParams> {

    override fun getActionCompletable(params: FriendRequestParams): Completable {
        return repository.sendFriendRequest(params.username, params.message)
    }

    class FriendRequestParams(val username: String, val message: String?)
}
